src/environment.c

```diff
@@ -0,0 +1,369 @@
+      warnx("invalid size for user key passed through %s environment variable: "
+            "expecting '%u', got '%i'",
+            USER_KEY_ENV_NAME, dst->d_len, size);
+      p = NULL;
+    }
+  }
+
+  if (p == NULL) {
+    fprintf(stderr,
+            "[*] Private key ('%s') initialized with default value: '%p'\n",
+            ENV_NAME, dst->d);
+    return false;
+  } else {
+    memcpy(dst->d, tmp, dst->d_len);
+    fprintf(stderr,
+            "[*] Private key ('%s') initialized from environment: '%s'\n",
+            ENV_NAME, p);
+    return true;
+  }
+}
+
+static void env_init_user_certificate(unsigned int index)
+{
+  env_user_certificate_t *certificate;
+  uint8_t pkey[65] = { 0 };
+  sys_os_endorsement_get_public_key(index, &pkey[0]);
+  uint8_t hash[32] = { 0 };
+  cx_sha256_t sha256;
+
+  switch (index) {
+  case 1:
+    certificate = &user_certificate_1;
+    break;
+  case 2:
+    certificate = &user_certificate_2;
+    break;
+  default:
+    THROW(EXCEPTION);
+    break;
+  }
+
+  sys_cx_sha256_init(&sha256);
+  sys_cx_hash((cx_hash_t *)&sha256, 0, (unsigned char *)"\xfe", 1, NULL, 0);
+  sys_cx_hash((cx_hash_t *)&sha256, CX_LAST, pkey, sizeof(pkey), hash,
+              sizeof(hash));
+  sys_cx_ecdsa_sign(&attestation_key, CX_RND_TRNG, CX_SHA256, hash,
+                    sizeof(hash), certificate->buffer, 6 + 33 * 2, NULL);
+  certificate->length = certificate->buffer[1] + 2;
+
+  fprintf(stderr, "[*] User certificate %u initialized (size: %u)\n", index,
+          certificate->length);
+}
+
+static void env_init_endorsement()
+{
+  bool attestation_changed =
+      env_init_user_hex_private_key(ATTESTATION_ENV_NAME, &attestation_key);
+  if (env_init_user_hex_private_key(USER_KEY_ENV_NAME, &user_private_key_1) ||
+      attestation_changed) {
+    env_init_user_certificate(1);
+  }
+  if (env_init_user_hex_private_key(USER_KEY_ENV_NAME, &user_private_key_2) ||
+      attestation_changed) {
+    env_init_user_certificate(2);
+  }
+}
+
+cx_ecfp_private_key_t *env_get_user_private_key(unsigned int index)
+{
+  switch (index) {
+  case 1:
+    return &user_private_key_1;
+    break;
+  case 2:
+    return &user_private_key_2;
+    break;
+  default:
+    THROW(EXCEPTION);
+    break;
+  }
+}
+
+env_user_certificate_t *env_get_user_certificate(unsigned int index)
+{
+  switch (index) {
+  case 1:
+    return &user_certificate_1;
+    break;
+  case 2:
+    return &user_certificate_2;
+    break;
+  default:
+    THROW(EXCEPTION);
+    break;
+  }
+}
+
+static void env_init_app_name_version()
+{
+  char *str;
+
+  str = getenv(APP_NAME_VERSION_ENV_NAME);
+  if (str == NULL) {
+    str = getenv(APP_NAME_VERSION_ENV_NAME_BKP);
+  }
+
+  if (str == NULL) {
+    warnx("using default app name & version");
+    fprintf(stderr, "[*] Default app name: '%s'\n", app_name.name);
+    fprintf(stderr, "[*] Default app version: '%s'\n", app_version.name);
+    return;
+  }
+
+  char *char_ptr = strchr(str, ':');
+  if (char_ptr == NULL) {
```

Comment:
There is a missing else line 317 to avoid the double warning?

Reply:
Hmm you mean `strchr(NULL, ':')` would return `NULL`? The early `return` line 315 should avoid that.

Reply:
Oh I missed that early return! Sorry